Data races detected

While runnning the tests with data race conditions on. Some were found.

    $ go run github.com/onsi/ginkgo/ginkgo -r --randomizeAllSpecs --randomizeSuites --failOnPending --cover --trace --race --progress

```
Services Tests
  Retrier
    Startable
      ✔ should start a service after the third try waiting between each
      ✔ should start and stop a service
    StartableWithContext
      ✔ should start and stop a service
    Startable
      ✔ should fail when a service takes too long to start
  Starter
    Stop
      ✔ should stop all services
  Retrier
    StartableWithContext
      ✔ should start a service after the third try
  Starter
    Start
      ✔ should fail starting a service that is not Startable
      ✔ should fallback stopping all started services after then last failed
      ✔ should interrupt starting services when load fails
  Retrier
    Startable
      ✔ should load a Configurable service
    StartableWithContext
      ✔ should fail when a service takes too long to start
==================
WARNING: DATA RACE
Read at 0x00c000438058 by goroutine 28:
  github.com/setare/services.(*Starter).Stop()
      /home/jota/projects/setare/services/starter.go:170 +0xc9
  github.com/setare/services.glob..func2.1.2.1()
      /home/jota/projects/setare/services/starter_test.go:162 +0x45

Previous write at 0x00c000438058 by goroutine 7:
  github.com/setare/services.(*Starter).Start()
      /home/jota/projects/setare/services/starter.go:46 +0xdf
  github.com/setare/services.glob..func2.1.2()
      /home/jota/projects/setare/services/starter_test.go:166 +0x461
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 28 (running) created at:
  github.com/setare/services.glob..func2.1.2()
      /home/jota/projects/setare/services/starter_test.go:158 +0x450
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:104 +0x337
==================
==================
WARNING: DATA RACE
Write at 0x00c00041e190 by goroutine 28:
  sync/atomic.CompareAndSwapInt32()
      /home/jota/.gvm/gos/go1.14.2/src/runtime/race_amd64.s:293 +0xb
  sync.(*Mutex).Lock()
      /home/jota/.gvm/gos/go1.14.2/src/sync/mutex.go:74 +0x49
  context.(*cancelCtx).cancel()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:395 +0x6c
  context.WithCancel.func1()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:235 +0x78
  github.com/setare/services.(*Starter).Stop()
      /home/jota/projects/setare/services/starter.go:171 +0x54c
  github.com/setare/services.glob..func2.1.2.1()
      /home/jota/projects/setare/services/starter_test.go:162 +0x45

Previous write at 0x00c00041e190 by goroutine 7:
  context.WithCancel()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:233 +0xa7
  github.com/setare/services.(*Starter).Start()
      /home/jota/projects/setare/services/starter.go:46 +0x72
  github.com/setare/services.glob..func2.1.2()
      /home/jota/projects/setare/services/starter_test.go:166 +0x461
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 28 (running) created at:
  github.com/setare/services.glob..func2.1.2()
      /home/jota/projects/setare/services/starter_test.go:158 +0x450
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:104 +0x337
==================
  Starter
    Start
      ✔ should interrupt starting services
    Stop
      ✔ should stop all services
  Retrier
    Startable
      ✔ should fail starting a service after reaching the tries limit
  Starter
    Start
      ✔ should fallback stopping all started services with context after then last failed
      ✔ should start a service
==================
WARNING: DATA RACE
Read at 0x00c0001b6f58 by goroutine 34:
  github.com/setare/services.(*Starter).Stop()
      /home/jota/projects/setare/services/starter.go:170 +0xc9
  github.com/setare/services.glob..func2.1.4.1()
      /home/jota/projects/setare/services/starter_test.go:239 +0x45

Previous write at 0x00c0001b6f58 by goroutine 7:
  github.com/setare/services.(*Starter).Start()
      /home/jota/projects/setare/services/starter.go:46 +0xdf
  github.com/setare/services.glob..func2.1.4()
      /home/jota/projects/setare/services/starter_test.go:243 +0x461
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 34 (running) created at:
  github.com/setare/services.glob..func2.1.4()
      /home/jota/projects/setare/services/starter_test.go:235 +0x450
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:104 +0x337
==================
==================
WARNING: DATA RACE
Write at 0x00c00033c1d0 by goroutine 34:
  sync/atomic.CompareAndSwapInt32()
      /home/jota/.gvm/gos/go1.14.2/src/runtime/race_amd64.s:293 +0xb
  sync.(*Mutex).Lock()
      /home/jota/.gvm/gos/go1.14.2/src/sync/mutex.go:74 +0x49
  context.(*cancelCtx).cancel()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:395 +0x6c
  context.WithCancel.func1()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:235 +0x78
  github.com/setare/services.(*Starter).Stop()
      /home/jota/projects/setare/services/starter.go:171 +0x54c
  github.com/setare/services.glob..func2.1.4.1()
      /home/jota/projects/setare/services/starter_test.go:239 +0x45

Previous write at 0x00c00033c1d0 by goroutine 7:
  context.WithCancel()
      /home/jota/.gvm/gos/go1.14.2/src/context/context.go:233 +0xa7
  github.com/setare/services.(*Starter).Start()
      /home/jota/projects/setare/services/starter.go:46 +0x72
  github.com/setare/services.glob..func2.1.4()
      /home/jota/projects/setare/services/starter_test.go:243 +0x461
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 34 (running) created at:
  github.com/setare/services.glob..func2.1.4()
      /home/jota/projects/setare/services/starter_test.go:235 +0x450
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/onsi/ginkgo@v1.12.0/ginkgo_dsl.go:226 +0x338
  github.com/novln/macchiato.RunSpecs()
      /home/jota/.gvm/pkgsets/go1.14.2/global/pkg/mod/github.com/novln/macchiato@v1.0.1/ginkgo.go:12 +0x2d8
  github.com/setare/services_test.TestServices()
      /home/jota/projects/setare/services/services_test.go:13 +0xf8
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /home/jota/.gvm/gos/go1.14.2/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:104 +0x337
==================
      ✔ should interrupt starting services with context
  Retrier
    StartableWithContext
      ✔ should fail starting a service after reaching the tries limit

Finished with success in 0.799 secs

 ✔ 18 test passed

--- FAIL: TestServices (0.80s)
    testing.go:906: race detected during execution of test
FAIL
coverage: 73.3% of statements

Ginkgo ran 1 suite in 4.681571365s
Test Suite Failed
```

--- a/starter.go
+++ b/starter.go
@@ -20,7 +20,9 @@ const (
 // for starting them when `Start` or `StartWithContext` is called.
 type Starter struct {
 	services             []Service
+	ctxMutex             sync.Mutex
 	startingMutex        sync.Mutex
+	stoppingMutex        sync.Mutex
 	startingContext      context.Context
 	servicesStarted      []Service
 	ctx                  context.Context
@@ -43,7 +45,9 @@ func NewStarter(services ...Service) *Starter {
 //
 // This method uses `StartWithContext`.
 func (s *Starter) Start() error {
+	s.ctxMutex.Lock()
 	s.ctx, s.cancelFunc = context.WithCancel(context.Background())
+	s.ctxMutex.Unlock()
 	return s.startWithContext(s.ctx)
 }
 
@@ -164,16 +168,29 @@ func (s *Starter) startWithContext(ctx context.Context) (err error) {
 // Stop will go through all started services, in the opposite order they were started, stopping one by one. If any,
 // failure is detected, the function will stop leaving some started services.
 func (s *Starter) Stop() error {
+	// Prevents two stops running at the sabe time
+	s.stoppingMutex.Lock()
+	defer s.stoppingMutex.Unlock()
+
+	s.ctxMutex.Lock()
+	if s.cancelFunc == nil {
+		// Already stopped
+		s.ctxMutex.Unlock()
+		return nil
+	}
+
+	s.cancelFunc() // Tries to cancel the starting proccess.
+	s.ctxMutex.Unlock()
+
 	if s.reporter != nil {
 		s.reporter.BeforeStop(nil)
 	}
-	if s.cancelFunc != nil {
-		s.cancelFunc() // Tries to cancel the starting proccess.
-	}
 
 	defer func() {
+		s.ctxMutex.Lock()
 		s.ctx = nil
 		s.cancelFunc = nil
+		s.ctxMutex.Unlock()
 
 		if s.reporter != nil {
 			s.reporter.AfterStop(nil, nil)

